### PR TITLE
Adapt image processing settings to JoomIMGtools

### PR DESCRIPTION
**Pull request by Elfangor93**
The sheer volume of configuration options in the configuration manager may be confusing for JoomGallery users.
This pull request is intended to provide a better structure and explanation of the options concerning image processing. It introduces a new tab under the tab "frontend settings" called "image processing". All options concerning image processing can be found now in one of these two tabs for image processing options:

- **General Settings/Image Processing** (This tab containing all image manipulation options which manipulate the images stored on the server [original, detail and thumb]. This options affect image upload and image manipulations triggered in the image manager.)
- **Frontend Settings/Image Processing** (This tab containing all image manipulation options which takes place on-the-fly, in that moment when the image is requested. Here images gets manipulated befor sending to the browser without saving them on the server.)

The options are arranged in both tabs regarding the image that is affected from this option (original image, detail image, thumbnail). This is intended to **_result in a more logical structure of the options_**. The description of the options has been changed at some places so that they are better aligned to each other and correctly reflect the behaviour of the JoomIMGtools class.

Additionally **_a tool to check and optimize_** your currently set options was added.
It is placed on the bottom of the `General Settings/Image Processing` tab. This tool allows you to upload an image of your choice and apply your settings to the uploaded image creating versions of thumbnail and detail images. You can generate up to two versions of each image (ideally generated with different settings). The tool allows you to directly compair the two images generated and to see the resulting differences.
This way, the image processing options can be optimized for any application without the need to add images to the database. The images uploaded and created by this tool are stored in the JoomGallery tmp-folder. Only one image at a time can exist in the tmp-folder. If a new image is generated or uploaded, the old one is overwritten.

### CODE REVIEW

**Comment by Elfangor93** [comment]
Here is a short video for better understanding how the **_tool to check and optimize_** the image processing configuration is designed to work. Please leave a comment if you would like any changes in this behaviour or design of the tool.

https://user-images.githubusercontent.com/39154009/112121018-0d1dc000-8bbf-11eb-81bc-1b6e488afe33.mp4

**Comment by eumel1602** [comment]
In the area where you can "check and optimize your currently set options" I compared two images ("compare images"). In the pop-up window at the bottom left, I change the "width" to a size (pixel) that is larger than the size of my screen (e.g. 1500 width). Then the popup window is so big that I can't get the value reset. I can only close it at the top right!

bild1

File: administrator/components/com_joomgallery/helpers/html/joomconfig.php [hunk 1]
```diff
@@ -129,4 +129,353 @@ public static function row($key, $type, $name, $info, $display = true, $attribs
...
+        </div>
+
+        <?php if($detail_exist || $thumb_exist): ?>
+          <div class="modal hide fade jviewport-width80 in" id="modalCompare" tabindex="-1" aria-labelledby="modalCompare_title" aria-hidden="true">
+            <div class="modal-header">
+              <button type="button" class="close novalidate" data-dismiss="modal" aria-label="Close">
+                <span aria-hidden="true">&times;</span>
+              </button>
+              <h3 id="modalCompare_title" thumb_img="<?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_COMPAREBOX_THUMBTITLE'); ?>" detail_img="<?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_COMPAREBOX_DETAILTITLE'); ?>">modal_title</h3>
+            </div>
```
Comment (MrMusic) [inline administrator/components/com_joomgallery/helpers/html/joomconfig.php:339]: Add
```
              <div class="center">
                <input id="modalImg_width" type="number" min="1" value="400">
                <span> px </span>
                <button id="modalImg_btn" class="btn btn-primary" onclick="setIMGwidth(event);"><?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_COMPAREBOX_APPLY'); ?></button>
              </div>
```
in modal header and remove modal-footer
to always be able to change the value of width.

File: administrator/components/com_joomgallery/helpers/html/joomconfig.php [hunk 2]
```diff
@@ -129,4 +129,353 @@ public static function row($key, $type, $name, $info, $display = true, $attribs
...
+                          <p><strong <?php echo JHTML::_('joomgallery.tip', JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_FRAMES_TXT'), JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_FRAMES_LBL'), true); ?>><?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_FRAMES_LBL');?>:</strong> <?php echo $info->detailL->frames;?></p>
+                          <p><strong <?php echo JHTML::_('joomgallery.tip', JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENTIME_TXT'), JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENTIME_LBL'), true); ?>><?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENTIME_LBL');?>:</strong> <?php echo round($info->detailL->processing_time,3);?> sec</p>
+                          <p><strong <?php echo JHTML::_('joomgallery.tip', JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENMEMORY_TXT'), JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENMEMORY_LBL'), true); ?>><?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_GENMEMORY_LBL');?>:</strong> <?php echo round(($info->detailL->used_memory) / pow(1000,2),2);?> MB</p>
+                        <?php else: ?>
```
Comment (MrMusic) [inline administrator/components/com_joomgallery/helpers/html/joomconfig.php:251]: Add
`                          <p><strong <?php echo JHTML::_('joomgallery.tip', JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_QUALITY_TXT'), JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_QUALITY_LBL'), true); ?>><?php echo JText::_('COM_JOOMGALLERY_CONFIG_GS_IP_TEST_QUALITY');?>:</strong> <?php echo ($info->detailL->used_quality);?></p>`
to show the quality parameter

**Review by MrMusic — COMMENTED** [review]
Show quality parameter.
The quality parameter is important for image quality so it should be saved/shown.
It's not finished but you should understand what I mean.

**Comment by MrMusic** [comment]
One more suggestion:
It would be nice if there was also something "to touch" at the top to be able to move the slider like at bottom.
image-compare-slider_klein

**Comment by Elfangor93** [comment]
>It would be nice if there was also something "to touch" at the top to be able to move the slider like at bottom.

Right now, this feature is realized with using pure HTML/CSS. The apperience of the slider is then done by the browser you are using. If we would like it more userfriendly, we need to use a javascript library to create the slider. This I wanted to avoid. Thats why it is like it is...

**Comment by MrMusic** [comment]
>Right now, this feature is realized with using pure HTML/CSS. The apperience of the slider is then done by the browser you are using. If we would like it more userfriendly, we need to use a javascript library to create the slider. This I wanted to avoid. Thats why it is like it is...

OK, if it's not easy to change we can leave it as it is.

**Comment by Elfangor93** [comment]
Attention not merge this PR I did a mistake with the rebase...

**Pull request merged**